File: frontend/src/types/finance.ts
export type FinanceAccountStatus = "active" | "suspended" | "closed";
export type FinanceTransactionType = "charge" | "payment" | "adjustment" | "refund";
export type FinanceTransactionSource = "booking" | "pos" | "order" | "manual" | "settlement";
export type FinanceSummaryPeriod = "day" | "week" | "month";

export interface FinanceAccountCustomerSummary {
  id: string;
  account_code: string;
  person_id: string;
}

export interface FinanceAccountSummary {
  id: string;
  club_id: string;
  account_customer_id: string;
  account_customer: FinanceAccountCustomerSummary;
  status: FinanceAccountStatus;
  balance: string;
  transaction_count: number;
}

export interface FinanceJournalEntry {
  id: string;
  club_id: string;
  account_id: string;
  amount: string;
  type: FinanceTransactionType;
  source: FinanceTransactionSource;
  reference_id: string | null;
  description: string;
  created_at: string;
  account_customer_code: string | null;
}

export interface FinanceClubJournal {
  entries: FinanceJournalEntry[];
  total_count: number;
}

export interface FinanceSummaryWindow {
  period: FinanceSummaryPeriod;
  date_from: string;
  date_to: string;
}

export interface FinanceRevenueSourceSummary {
  source: FinanceTransactionSource;
  total_revenue: string;
  charge_count: number;
}

export interface FinanceRevenuePeriodSummary extends FinanceSummaryWindow {
  total_revenue: string;
  operational_revenue: string;
  charge_count: number;
  by_source: FinanceRevenueSourceSummary[];
}

export interface FinanceRevenueSummary {
  timezone: string;
  reference_datetime: string;
  day: FinanceRevenuePeriodSummary;
  week: FinanceRevenuePeriodSummary;
  month: FinanceRevenuePeriodSummary;
}

export interface FinanceTransactionVolumeTypeSummary {
  type: FinanceTransactionType;
  transaction_count: number;
  total_absolute_amount: string;
}

export interface FinanceTransactionVolumePeriodSummary extends FinanceSummaryWindow {
  total_transaction_count: number;
  by_type: FinanceTransactionVolumeTypeSummary[];
}

export interface FinanceTransactionVolumeSummary {
  timezone: string;
  reference_datetime: string;
  day: FinanceTransactionVolumePeriodSummary;
  week: FinanceTransactionVolumePeriodSummary;
  month: FinanceTransactionVolumePeriodSummary;
}

export interface FinanceOutstandingSummary {
  total_accounts: number;
  accounts_in_arrears: number;
  accounts_in_credit: number;
  accounts_settled: number;
  total_outstanding_amount: string;
  unpaid_order_postings_count: number;
  unpaid_order_postings_amount: string;
  pending_items_count: number;
}

export interface FinanceLedgerEntry extends FinanceJournalEntry {
  running_balance: string;
}

export interface FinanceAccountLedger {
  account_id: string;
  club_id: string;
  account_customer_id: string;
  status: FinanceAccountStatus;
  balance: string;
  transactions: FinanceLedgerEntry[];
}

export type FinanceExportProfile = "journal_basic";
export type FinanceExportBatchStatus = "draft" | "generated" | "exported" | "void";

export interface FinanceExportBatchPreviewRow {
  entry_date: string;
  transaction_id: string;
  account_customer_code: string | null;
  transaction_type: string;
  source: string;
  reference_id: string | null;
  description: string;
  amount: string;
  debit_amount: string;
  credit_amount: string;
}

export interface FinanceExportBatchSummary {
  id: string;
  club_id: string;
  export_profile: FinanceExportProfile;
  date_from: string;
  date_to: string;
  status: FinanceExportBatchStatus;
  created_by_person_id: string;
  generated_at: string;
  file_name: string;
  content_hash: string;
  transaction_count: number;
  total_debits: string;
  total_credits: string;
  metadata_json: {
    selection_timezone?: string;
    selection_window?: {
      date_from?: string;
      date_to?: string;
    };
    source_counts?: Record<string, number>;
    transaction_type_counts?: Record<string, number>;
  };
}

export interface FinanceExportBatchDetail extends FinanceExportBatchSummary {
  rows: FinanceExportBatchPreviewRow[];
}

export interface FinanceExportBatchCreateInput {
  export_profile: FinanceExportProfile;
  date_from: string;
  date_to: string;
}

export interface FinanceExportBatchCreateResult {
  created: boolean;
  batch: FinanceExportBatchDetail;
}

export interface FinanceExportBatchListResponse {
  batches: FinanceExportBatchSummary[];
  total_count: number;
}

export interface FinanceExportBatchVoidResult {
  void_applied: boolean;
  batch: FinanceExportBatchDetail;
}

export type FinanceTargetSystem = "generic_journal" | "pastel_like" | "sage_like";

export interface AccountingExportProfileTransactionMapping {
  debit_account_code: string;
  credit_account_code: string;
  description_prefix: string;
}

export interface AccountingExportProfileMappingConfig {
  reference_prefix: string;
  fallback_customer_code: string;
  transaction_mappings: Partial<Record<FinanceTransactionType, AccountingExportProfileTransactionMapping>>;
}

export interface AccountingExportProfile {
  id: string;
  club_id: string;
  code: string;
  name: string;
  target_system: string;
  is_active: boolean;
  mapping_config: AccountingExportProfileMappingConfig;
  created_by_person_id: string;
  created_at: string;
  updated_at: string;
}

export interface AccountingExportProfileInput {
  code: string;
  name: string;
  target_system: string;
  is_active: boolean;
  mapping_config: AccountingExportProfileMappingConfig;
}

export interface AccountingExportProfileListResponse {
  profiles: AccountingExportProfile[];
  total_count: number;
}

export interface AccountingMappedExportPreviewRow {
  date: string;
  reference: string;
  description: string;
  debit_account_code: string;
  credit_account_code: string;
  amount: string;
  customer_account_code: string;
  source_type: string;
}

export interface AccountingMappedExportValidationError {
  code: string;
  message: string;
  row_index: number | null;
  field: string | null;
}

export interface AccountingMappedExportPreview {
  source_batch_id: string;
  source_export_profile: FinanceExportProfile;
  accounting_profile_id: string;
  accounting_profile_code: string;
  accounting_profile_name: string;
  target_system: string;
  generated_at: string;
  file_name: string;
  content_hash: string;
  row_count: number;
  download_ready: boolean;
  metadata_json: {
    output_mode?: string;
    source_batch_content_hash?: string;
    source_batch_file_name?: string;
    column_order?: string[];
  };
  validation_errors: AccountingMappedExportValidationError[];
  rows: AccountingMappedExportPreviewRow[];
}

File: backend/app/schemas/finance.py
from __future__ import annotations

import uuid
from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

from app.models import (
    FinanceAccountStatus,
    FinanceExportBatchStatus,
    FinanceExportProfile,
    FinanceTransactionSource,
    FinanceTransactionType,
)


class FinanceTransactionCreateRequest(BaseModel):
    account_id: uuid.UUID
    amount: Decimal
    type: FinanceTransactionType
    source: FinanceTransactionSource
    reference_id: uuid.UUID | None = None
    description: str = Field(min_length=1, max_length=255)

    @field_validator("amount")
    @classmethod
    def validate_non_zero_amount(cls, value: Decimal) -> Decimal:
        if value == 0:
            raise ValueError("amount must be non-zero")
        return value

    @field_validator("description")
    @classmethod
    def normalize_description(cls, value: str) -> str:
        return value.strip()


class FinanceTransactionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    club_id: uuid.UUID
    account_id: uuid.UUID
    amount: Decimal
    type: FinanceTransactionType
    source: FinanceTransactionSource
    reference_id: uuid.UUID | None = None
    description: str
    created_at: datetime


class FinanceTransactionCreateResult(BaseModel):
    transaction: FinanceTransactionResponse
    balance: Decimal


class FinanceLedgerEntryResponse(FinanceTransactionResponse):
    running_balance: Decimal


class FinanceAccountLedgerResponse(BaseModel):
    account_id: uuid.UUID
    club_id: uuid.UUID
    account_customer_id: uuid.UUID
    status: FinanceAccountStatus
    balance: Decimal
    transactions: list[FinanceLedgerEntryResponse]


class FinanceAccountCustomerSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    account_code: str
    person_id: uuid.UUID


class FinanceAccountSummaryResponse(BaseModel):
    id: uuid.UUID
    club_id: uuid.UUID
    account_customer_id: uuid.UUID
    account_customer: FinanceAccountCustomerSummary
    status: FinanceAccountStatus
    balance: Decimal
    transaction_count: int


class FinanceJournalEntryResponse(FinanceTransactionResponse):
    account_customer_code: str | None


class FinanceClubJournalResponse(BaseModel):
    entries: list[FinanceJournalEntryResponse]
    total_count: int


class FinanceExportBatchPreviewRow(BaseModel):
    entry_date: str
    transaction_id: str
    account_customer_code: str | None
    transaction_type: str
    source: str
    reference_id: str | None
    description: str
    amount: str
    debit_amount: str
    credit_amount: str


class FinanceExportBatchCreateRequest(BaseModel):
    export_profile: FinanceExportProfile
    date_from: date
    date_to: date

    @model_validator(mode="after")
    def validate_range(self) -> "FinanceExportBatchCreateRequest":
        if self.date_to < self.date_from:
            raise ValueError("date_to must be on or after date_from")
        return self


class FinanceExportBatchSummaryResponse(BaseModel):
    id: uuid.UUID
    club_id: uuid.UUID
    export_profile: FinanceExportProfile
    date_from: date
    date_to: date
    status: FinanceExportBatchStatus
    created_by_person_id: uuid.UUID
    generated_at: datetime
    file_name: str
    content_hash: str
    transaction_count: int
    total_debits: Decimal
    total_credits: Decimal
    metadata_json: dict[str, object]


class FinanceExportBatchDetailResponse(FinanceExportBatchSummaryResponse):
    rows: list[FinanceExportBatchPreviewRow]


class FinanceExportBatchCreateResult(BaseModel):
    created: bool
    batch: FinanceExportBatchDetailResponse


class FinanceExportBatchListResponse(BaseModel):
    batches: list[FinanceExportBatchSummaryResponse]
    total_count: int


class FinanceExportBatchVoidResult(BaseModel):
    void_applied: bool
    batch: FinanceExportBatchDetailResponse


class FinanceExportBatchDownloadResult(BaseModel):
    file_name: str
    content: str
